feat(zcoin): ARRR WASM implementation

borngraced opened the pull request:
This PR integrates zcash wallet db in WASM. All changes has been tested really well.

## Tasks Completed in this Pull Request:

- [x] Implemented all `blocksdb` I/O functions for `WASM` including unit tests
- [x]  Refactored the `light_wallet_db_synch_loop` function to be fully asynchronous and removed all uses of `block_on` and `block_in_place`.
- [x] Renamed the `scan_blocks` function to `scan_validate_and_update_blocks` and refactored it into an asynchronous function.
- [x] Refactored the `validate_chain` and `scan_cached_block` functions to be fully asynchronous. Simplified the function processes for better manageability.
- [x]  Implementation of walletdb + unit tests for wallets I/O functions for WASM.
- [x] Implementation of `zcash_params` an unit tests using  `indexeddb`
- [x] Implemented tonic transport layer for `ARRR` that works almost same way with the current native transport and also support streaming data
- [x] completed `ARRR` web integration with changes to `z_coin` mod.  

## Deps Additions
- [base64](https://docs.rs/base64/0.21.2/base64/index.html)(0.21.7) - Base64 encoding and decoding functionality used for WASM tonic client
- [tower-service](https://docs.rs/tower-service/latest/tower_service/)(0.3.2) - Defines the Service trait for building composable network services used for WASM tonic client.
- [blake2b_simd](https://crates.io/crates/blake2b_simd/0.5.10)(0.5.10) - Rust implementation of the BLAKE2b cryptographic hash function for working with `z_coin transactions`
- [ff](https://crates.io/crates/ff/0.8.0)(0.8.0) - Finite field arithmetic operations for working with `z_coin transactions`.
- [jubjub](https://crates.io/crates/jubjub/0.5.1)(0.5.1) - ubjub elliptic curve cryptography for working with `z_coin transactions`.
- [time](https://crates.io/crates/time/0.3.20)(0.3.20, features = ["wasm-bindgen"]) - time library for Rust, with WebAssembly support for working with `z_coin transactions`.

## Deps Updated

- proc-macro2 [1.0.63](https://crates.io/crates/proc-macro2/1.0.63) -> [1.0.69](https://crates.io/crates/proc-macro2/1.0.69)
- quote [1.0.28](https://crates.io/crates/quote/1.0.28) -> [1.0.33](https://crates.io/crates/quote/1.0.33)
- memoffset 0[.5.4](https://crates.io/crates/memoffset/0.5.4) -> [0.5.6](https://crates.io/crates/memoffset/0.5.6)
- syn 2.0.23 -> 2.0.38
- serde_derive [1.0.164](https://crates.io/crates/syn/1.0.64) -> [1.0.189](https://crates.io/crates/syn/1.0.89)
- serde_urlencoded [0.7.0](https://crates.io/crates/serde_urlencoded/0.7.0) -> [0.7.1](https://crates.io/crates/serde_urlencoded/0.7.1)
- time [0.3.11](https://crates.io/crates/time/0.3.11) -> [0.3.20](https://crates.io/crates/time/0.3.20)
- time-macros [0.2.4](https://crates.io/crates/time-macros/0.2.4) -> [0.2.8](https://crates.io/crates/time-macros/0.2.8)

## NOTE
Some `walletdb` wasm tests are commented out because of long running time.

onur-ozkan commented on mm2src/mm2_db/src/indexed_db/indexed_db.rs:
```diff
@@ -306,6 +306,15 @@ pub enum AddOrIgnoreResult {
     ExistAlready(ItemId),
 }
 
+impl AddOrIgnoreResult {
+    pub fn item_id(self) -> ItemId {
+        match self {
+            AddOrIgnoreResult::Added(item) => item,
+            AddOrIgnoreResult::ExistAlready(item) => item,
+        }
+    }
+}
+
```
Can you implement `From` trait instead of this custom conversion?

onur-ozkan commented on mm2src/coins/z_coin/z_rpc.rs:
```diff
@@ -135,17 +119,18 @@ impl ZRpcOps for LightRpcClient {
             .map_to_mm(UpdateBlocksCacheErr::GrpcError)?
             .into_inner();
-        // without Pin method get_mut is not found in current scope
+        //         without Pin method get_mut is not found in current scope
```
Can you revert this whitespace?

onur-ozkan commented on mm2src/coins/z_coin/z_rpc.rs:
```diff
@@ -257,8 +243,12 @@ impl ZRpcOps for NativeClient {
                 vtx: compact_txs,
             };
-            on_block(compact_block)?;
+            db.insert_block(compact_block.height as u32, compact_block.encode_to_vec())
```
Instead of doing direct cast from u64 to u32, can you handle the overflow?

shamardy replied:
This is still not fixed.

onur-ozkan commented on mm2src/coins/z_coin/z_rpc.rs:
```diff
@@ -719,7 +671,9 @@ async fn light_wallet_db_sync_loop(mut sync_handle: SaplingSyncLoopHandle, mut c
 
         if let Some(tx_id) = sync_handle.watch_for_tx {
-            if !block_in_place(|| is_tx_imported(sync_handle.wallet_db.db.lock().sql_conn(), tx_id)) {
+            let walletdb = &sync_handle.wallet_db;
+            let is_tx_imported = walletdb.is_tx_imported(tx_id).await;
```
```suggestion
            let is_tx_imported = &sync_handle.wallet_db.is_tx_imported(tx_id).await;
```

onur-ozkan replied:
I thought it already seems understandable from the suggestion. Creating another variable doesn't seem to help in any way. The used field already points what its referring to.

onur-ozkan commented on mm2src/coins/z_coin/storage/blockdb/blockdb_idb_storage.rs:
```diff
@@ -0,0 +1,252 @@
+}
+
+impl TableSignature for BlockDbTable {
+    fn table_name() -> &'static str { "compactblocks" }
```
Updating `TableSignature` trait would be better in favor of replacing `fn table_name() -> &static str` to `const TABLE_NAME: &str';`

onur-ozkan replied:
Yes, as it is quite easy change(just replacing bunch of lines)

borngraced replied:
done

onur-ozkan commented on mm2src/coins/z_coin/storage/blockdb/blockdb_idb_storage.rs:
```diff
@@ -0,0 +1,252 @@
+    }
+}
+
+pub struct BlockDbInner {
+    pub inner: IndexedDb,
+}
```
Let's make this tuple as it will always have single field.

onur-ozkan commented on mm2src/coins/z_coin/storage/blockdb/blockdb_idb_storage.rs:
```diff
@@ -0,0 +1,252 @@
+        let height_to_remove_from = height + 1;
+        for i in height_to_remove_from..get_latest_block {
+            let index_keys = MultiIndex::new(BlockDbTable::TICKER_HEIGHT_INDEX)
+                .with_value(&ticker)?
```
No need to clone the ticker, can use `&self.ticker` directly

onur-ozkan commented on mm2src/coins/z_coin/storage/blockdb/blockdb_idb_storage.rs:
```diff
@@ -0,0 +1,252 @@
+                &BlockDbTable {
+                    height,
+                    data: cb_bytes,
+                    ticker: ticker.clone(),
```
No need to clone the ticker twice, `self.ticker.clone()` can be used directly

onur-ozkan commented on mm2src/coins/z_coin/storage/blockdb/blockdb_idb_storage.rs:
```diff
@@ -0,0 +1,252 @@
+        // Fetch CompactBlocks block_db are needed for scanning.
+        let mut maybe_blocks = block_db
+            .cursor_builder()
+            .only("ticker", ticker.clone())?
```
Same here and on other multiple spots in this module.

onur-ozkan commented on mm2src/coins/z_coin/storage/blockdb/blockdb_sql_storage.rs:
```diff
@@ -0,0 +1,208 @@
+        validate_from: Option<(BlockHeight, BlockHash)>,
+        limit: Option<u32>,
+    ) -> MmResult<(), ZcoinStorageError> {
+        let ticker = self.ticker.to_owned();
```
This isn't needed

onur-ozkan reviewed (changes requested):
Great wor! Some notes from the first review iteration:

laruh commented on mm2src/coins/z_coin/storage/blockdb/blockdb_idb_storage.rs:
Please add doc comments for pub items

borngraced replied:
noted.

laruh commented on mm2src/coins/z_coin/storage/blockdb/blockdb_sql_storage.rs:
```diff
@@ -0,0 +1,208 @@
+
+impl BlockDbImpl {
+    #[cfg(all(not(test)))]
+    pub async fn new(_ctx: MmArc, ticker: String, path: Option<impl AsRef<Path>>) -> MmResult<Self, ZcoinStorageError> {
```
Please add doc comment

laruh commented on mm2src/coins/z_coin/storage/walletdb/mod.rs:
```diff
@@ -1,86 +1,21 @@
-}
-
 #[derive(Clone)]
 pub struct WalletDbShared {
```
doc comments

laruh commented on mm2src/coins/z_coin/storage/walletdb/mod.rs:
```diff
@@ -1,86 +1,21 @@
 cfg_wasm32!(
-    mod wallet_idb;
-    use wallet_idb::WalletDbInner;
+    pub mod wallet_idb_storage;
```
doc comments

laruh commented on mm2src/coins/z_coin/storage/walletdb/wallet_idb_storage.rs:
Please add doc comments for `pub` items

laruh commented on mm2src/coins/z_coin/z_coin_errors.rs:
```diff
@@ -279,3 +280,166 @@ pub enum SpendableNotesError {
 
 #[derive(Debug, Display)]
 pub enum ZCoinBalanceError {}
+
+#[derive(Debug, Display)]
+pub enum ValidateBlocksError {
```
Doc  comments

laruh commented on mm2src/coins/z_coin/z_coin_errors.rs:
```diff
@@ -279,3 +280,166 @@ pub enum SpendableNotesError {
+}
+
+impl ValidateBlocksError {
+    pub fn prev_hash_mismatch(height: BlockHeight) -> ValidateBlocksError {
```
doc comment

laruh commented on mm2src/coins/z_coin/z_coin_errors.rs:
```diff
@@ -279,3 +280,166 @@ pub enum SpendableNotesError {
+        }
+    }
+
+    pub fn block_height_discontinuity(height: BlockHeight, found: BlockHeight) -> ValidateBlocksError {
```
doc comment

laruh commented on mm2src/coins/z_coin/z_coin_errors.rs:
```diff
@@ -279,3 +280,166 @@ pub enum SpendableNotesError {
+
+#[allow(unused)]
+#[derive(Debug, Display)]
+pub enum ZcoinStorageError {
```
doc comment

laruh commented on mm2src/coins/z_coin/z_rpc.rs:
```diff
@@ -55,22 +43,16 @@ cfg_native!(
-pub type OnCompactBlockFn<'a> = dyn FnMut(String) -> Result<(), MmError<UpdateBlocksCacheErr>> + Send + 'a;
-
 #[async_trait]
 pub trait ZRpcOps {
```
left note about doc coms in this [PR](https://github.com/KomodoPlatform/komodo-defi-framework/pull/1922), I believe it will be merged first. Just a note.

borngraced replied:
yep, seen thanks @laruh

laruh commented on mm2src/mm2_db/src/indexed_db/indexed_db.rs:
```diff
@@ -306,6 +306,15 @@ pub enum AddOrIgnoreResult {
 }
 
+impl AddOrIgnoreResult {
+    pub fn item_id(self) -> ItemId {
```
doc comment

laruh reviewed (changes requested):
Huge work!
Please add doc comments for pub items.
Yeah we have much more pub places, if you see some pub struct, enum, fn, trait whatever, and you feel you can provide its description, please do it in next sprints.

onur-ozkan commented on mm2src/coins/Cargo.toml:
```diff
@@ -117,6 +118,8 @@ spl-token = { version = "3", optional = true }
 spl-associated-token-account = { version = "1", optional = true }
 
 [target.'cfg(target_arch = "wasm32")'.dependencies]
+ff = "0.8"
```
Can be removed
```suggestion
```

borngraced replied:
done

onur-ozkan replied:
They are different dependencies?

onur-ozkan replied:
Aah, I see, `from_raw` is the trait from `ff`.

onur-ozkan commented on mm2src/coins/z_coin/storage.rs:
```diff
@@ -1,5 +1,215 @@
+    pub fn inner(&self) -> DataConnStmtCacheWasm { self.clone().cache }
+}
+
+#[allow(unused)]
+pub struct CompactBlockRow {
```
```suggestion
pub struct CompactBlockRow {
```

borngraced replied:
done

onur-ozkan commented on mm2src/coins/z_coin/storage.rs:
```diff
@@ -1,5 +1,215 @@
+/// provides accurate block information as of the time it was requested.
+///
+pub async fn validate_chain(
+    block: CompactBlock,
+    prev_height: &mut BlockHeight,
```
Q: Is there a way to reach `prev_height` from `block` ?

borngraced replied:
nop actually.

onur-ozkan commented on mm2src/coins/z_coin/storage.rs:
```diff
@@ -1,5 +1,215 @@
+    }
+
+    let block_hash = BlockHash::from_slice(&block.hash);
+    let block_time = block.time;
```
I highly recommend avoid doing this as much as possible :)

onur-ozkan commented on mm2src/coins/z_coin/storage.rs:
```diff
@@ -1,5 +1,215 @@
+    #[cfg(debug_assertions)]
+    {
+        let cur_root = tree.root();
+        for row in &witnesses {
+            if row.1.root() != cur_root {
+                return Err(Error::InvalidWitnessAnchor(row.0, current_height).into());
+            }
+        }
```
much simpler I believe with less nested block

```suggestion
        if witnesses.iter().any(|row| row.1.root() != cur_root) {
            return Err(Error::InvalidWitnessAnchor(row.0, current_height).into());
        }
```

borngraced replied:
thanks, done

onur-ozkan commented on mm2src/coins/z_coin/z_coin_errors.rs:
```diff
@@ -279,3 +280,170 @@ pub enum SpendableNotesError {
+
+/// The `ZcoinStorageError` enum encapsulates different types of errors that may occur
+/// when interacting with storage operations specific to the Zcoin blockchain.
+#[allow(unused)]
+#[derive(Debug, Display)]
+pub enum ZcoinStorageError {
```
```suggestion
#[derive(Debug, Display)]
pub enum ZcoinStorageError {
```

onur-ozkan commented on mm2src/coins/z_coin/storage/walletdb/wallet_sql_storage.rs:
```diff
@@ -0,0 +1,88 @@
+        let db = self.db.inner();
+        async_blocking(move || {
+            let conn = db.lock().unwrap();
+            const QUERY: &str = "SELECT id_tx FROM transactions WHERE txid = ?1;";
```
You can use `SELECT EXISTS` here.

borngraced replied:
done

onur-ozkan reviewed (changes requested):
Finalized reviewing the PR, except "mm2src/coins/z_coin/storage/walletdb/wallet_idb_storage.rs" module. Please ping me once it's ready :)

Here are some other notes:

shamardy commented on mm2src/coins/Cargo.toml:
```diff
@@ -103,9 +103,10 @@ uuid = { version = "1.2.2", features = ["fast-rng", "serde", "v4"] }
-zcash_client_backend = { git = "https://github.com/KomodoPlatform/librustzcash.git", tag = "k-1.3.0" }
-zcash_primitives = { features = ["transparent-inputs"], git = "https://github.com/KomodoPlatform/librustzcash.git", tag = "k-1.3.0" }
-zcash_proofs = { git = "https://github.com/KomodoPlatform/librustzcash.git", tag = "k-1.3.0" }
+zcash_client_backend = { git = "https://github.com/KomodoPlatform/librustzcash.git", rev = "8f4b5e1" }
+zcash_extras = { git = "https://github.com/KomodoPlatform/librustzcash.git", rev = "8f4b5e1" }
+zcash_primitives = { features = ["transparent-inputs"], git = "https://github.com/KomodoPlatform/librustzcash.git", rev = "8f4b5e1" }
+zcash_proofs = { git = "https://github.com/KomodoPlatform/librustzcash.git", rev = "8f4b5e1" }
```
Just a reminder that the librustzcash PR https://github.com/KomodoPlatform/librustzcash/pull/8 needs to be reviewed, merged and released before merging this PR.

shamardy replied:
@borngraced can you please use the latest librustzcash release that includes async WalletDB https://github.com/KomodoPlatform/librustzcash/releases/tag/k-1.4.0
Please also resolve conflicts. I will do a last review of this PR after this.

borngraced replied:
done @shamardy

onur-ozkan commented:
Can you resolve the conflicting files?

borngraced commented:
> Can you resolve the conflicting files?

done @onur-ozkan

onur-ozkan commented on mm2src/coins/qrc20/qrc20_tests.rs:
```diff
@@ -1,12 +1,14 @@
 use itertools::Itertools;
 use mm2_core::mm_ctx::MmCtxBuilder;
 use mm2_number::bigdecimal::Zero;
+#[cfg(not(target_arch = "wasm32"))]
```
you can use cfg_native macro instead of declaring this multiple times

onur-ozkan commented on mm2src/coins/z_coin/storage.rs:
```diff
@@ -1,5 +1,204 @@
+
+#[cfg(target_arch = "wasm32")]
+use walletdb::wasm::storage::DataConnStmtCacheWasm;
+#[cfg(debug_assertions)]
+use zcash_client_backend::data_api::error::Error;
```
This will never compile.

https://github.com/KomodoPlatform/komodo-defi-framework/blob/79f620559a635704959d3e88067f7ffe5ee30f1f/Cargo.toml#L51-L70

borngraced replied:
thanks. removed [cfg(debug_assertions)]

onur-ozkan commented on mm2src/coins/z_coin/storage/walletdb/wallet_sql_storage.rs:
```diff
@@ -0,0 +1,118 @@
+        async_blocking(move || {
+            let conn = db.lock().unwrap();
+            const QUERY: &str = "SELECT EXISTS (SELECT 1 FROM transactions WHERE txid = ?1);";
+            match query_single_row(conn.sql_conn(), QUERY, [tx_id.0.to_vec()], |row| row.get::<_, i64>(0)) {
+                Ok(Some(_)) => Ok(true),
+                Ok(None) | Err(_) => Ok(false),
+            }
```
Something like

```rs
query_single_row(conn.sql_conn(), QUERY, [tx_id.0.to_vec()], |row| row.get::<_, i64>(0))
    .map(|v| v.is_some())
    .unwrap_or_default(),
```

should do the same thing.

onur-ozkan commented on mm2src/coins/z_coin/storage.rs:
```diff
@@ -1,5 +1,204 @@
+            *prev_height + 1,
+            current_height,
+        ))
+    } else {
+        match prev_hash {
+            None => Ok(()),
+            Some(ref h) if h == &block.prev_hash() => Ok(()),
+            Some(_) => Err(ValidateBlocksError::prev_hash_mismatch(current_height)),
+        }
+    }?;
```
shorter and less nested way:

```suggestion
    } else if prev_hash.is_none() || (prev_hash.as_ref() == Some(&block.prev_hash())) {
        Ok(())
    } else {
        Err(ValidateBlocksError::prev_hash_mismatch(current_height))
    }?;
```

borngraced replied:
done

onur-ozkan reviewed (changes requested):
First review iteration for the recent changes:

onur-ozkan commented:
By the way lock file isn't synced. I run `cargo check`, and got this diff:

```diff
diff --git a/Cargo.lock b/Cargo.lock
index 1970f36f5..ce70dcd2c 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -9414,7 +9414,7 @@ dependencies = [
  "bls12_381",
  "bs58 0.4.0",
  "ff 0.8.0",
- "futures 0.3.15",
+ "futures 0.3.28",
  "group 0.8.0",
  "hex 0.4.3",
  "jubjub",
@@ -9438,7 +9438,7 @@ dependencies = [
  "bech32",
  "bs58 0.4.0",
  "ff 0.8.0",
- "futures 0.3.15",
+ "futures 0.3.28",
  "group 0.8.0",
  "jubjub",
  "libsqlite3-sys",
```

laruh reviewed (changes requested):
Huge work! I have couple notes for now

laruh commented on mm2src/coins/z_coin/storage/walletdb/wasm/storage.rs:
```diff
@@ -0,0 +1,1474 @@
+            .next()
+            .await?;
+        if maybe_min_account.is_some() {
+            return MmError::err(ZcoinStorageError::TableNotEmpty(
+                "Account table is not empty".to_string(),
```
you have places, where you repeat 
```rust
ZcoinStorageError::TableNotEmpty( "Account table is not empty".to_string())
```
please use
```rust
#[display(fmt = "Account table is not empty")]
TableNotEmpty,
 ```

borngraced replied:
thanks for the suggestion, the thing this error variant might be use for other tables aside `AccountTable` later

laruh replied:
Hmm then I suggest to use
```rust
#[display(fmt = "{table_name} table is not empty")]
    TableNotEmpty { 
        table_name: String,
    },
```    
to not repeat "table is not empty" every time

borngraced replied:
great thanks!

laruh commented on mm2src/coins/z_coin/storage/walletdb/wasm/storage.rs:
```diff
@@ -0,0 +1,1474 @@
+        Ok(db)
+    }
+
+    #[allow(unused)]
```
I see you dont need `#[allow(unused)]` anymore

laruh commented on mm2src/coins/z_coin/storage/walletdb/wasm/storage.rs:
```diff
@@ -0,0 +1,1474 @@
+                .with_value(account_int)?;
+
+            walletdb_account_table
+                .replace_item_by_unique_multi_index(index_keys, &account)
```
Question: why do you use replace item and not add? just in case?
There is Insert into command for this functionality in sql.

borngraced replied:
actually got this advice from Sergey, as this ensures we will never duplicate value.. literally an insert too

onur-ozkan commented on mm2src/coins/z_coin/storage/walletdb/wasm/storage.rs:
```diff
@@ -0,0 +1,1494 @@
+        time: u32,
+        sapling_tree: &[u8],
+    ) -> MmResult<(), ZcoinStorageError> {
+        let ticker = self.ticker.clone();
+        let locked_db = self.lock_db().await?;
+        let db_transaction = locked_db.get_inner().transaction().await?;
+        let walletdb_account_table = db_transaction.table::<WalletDbBlocksTable>().await?;
+
+        // check if account exists
+        let maybe_min_account = walletdb_account_table
+            .cursor_builder()
+            .only("ticker", &ticker)?
+            .bound("height", 0u32, u32::MAX)
+            .open_cursor(WalletDbBlocksTable::TICKER_HEIGHT_INDEX)
+            .await?
+            .next()
+            .await?;
+        if maybe_min_account.is_some() {
+            return MmError::err(ZcoinStorageError::TableNotEmpty(
+                "Account table is not empty".to_string(),
+            ));
+        }
+
+        let block = WalletDbBlocksTable {
+            height: u32::from(height),
+            hash: hash.0.to_vec(),
+            time,
+            sapling_tree: sapling_tree.to_vec(),
+            ticker: ticker.clone(),
+        };
+        let walletdb_blocks_table = db_transaction.table::<WalletDbBlocksTable>().await?;
+        let height = u32::from(height);
+        let index_keys = MultiIndex::new(WalletDbBlocksTable::TICKER_HEIGHT_INDEX)
+            .with_value(&ticker)?
```
you can avoid double cloning with the following patch:

```diff
diff --git a/mm2src/coins/z_coin/storage/walletdb/wasm/storage.rs b/mm2src/coins/z_coin/storage/walletdb/wasm/storage.rs
index 03dd2df58..68b323e0a 100644
--- a/mm2src/coins/z_coin/storage/walletdb/wasm/storage.rs
+++ b/mm2src/coins/z_coin/storage/walletdb/wasm/storage.rs
@@ -228,7 +228,6 @@ impl<'a> WalletIndexedDb {
         time: u32,
         sapling_tree: &[u8],
     ) -> MmResult<(), ZcoinStorageError> {
-        let ticker = self.ticker.clone();
         let locked_db = self.lock_db().await?;
         let db_transaction = locked_db.get_inner().transaction().await?;
         let walletdb_account_table = db_transaction.table::<WalletDbBlocksTable>().await?;
@@ -236,7 +235,7 @@ impl<'a> WalletIndexedDb {
         // check if account exists
         let maybe_min_account = walletdb_account_table
             .cursor_builder()
-            .only("ticker", &ticker)?
+            .only("ticker", &self.ticker)?
             .bound("height", 0u32, u32::MAX)
             .open_cursor(WalletDbBlocksTable::TICKER_HEIGHT_INDEX)
             .await?
@@ -253,12 +252,12 @@ impl<'a> WalletIndexedDb {
             hash: hash.0.to_vec(),
             time,
             sapling_tree: sapling_tree.to_vec(),
-            ticker: ticker.clone(),
+            ticker: self.ticker.clone(),
         };
         let walletdb_blocks_table = db_transaction.table::<WalletDbBlocksTable>().await?;
         let height = u32::from(height);
         let index_keys = MultiIndex::new(WalletDbBlocksTable::TICKER_HEIGHT_INDEX)
-            .with_value(&ticker)?
+            .with_value(&self.ticker)?
             .with_value(num_to_bigint!(height)?)?;
         walletdb_blocks_table
             .replace_item_by_unique_multi_index(index_keys, &block)
```

onur-ozkan commented on mm2src/coins/z_coin/storage/walletdb/wasm/storage.rs:
```diff
@@ -0,0 +1,1494 @@
+            hash: hash.to_vec(),
+            time: block_time,
+            sapling_tree: encoded_tree,
+            ticker: ticker.clone(),
```
Similar to above, we are doing double cloning here

onur-ozkan commented on mm2src/coins/z_coin/storage/walletdb/wasm/storage.rs:
```diff
@@ -0,0 +1,1494 @@
+    }
+
+    pub async fn get_balance(&self, account: AccountId) -> MmResult<Amount, ZcoinStorageError> {
+        let ticker = self.ticker.clone();
```
We don't need to clone this, as it's only passed by reference. In this module we have too many extra cloning like this, I don't want to point them all one by one, so I leave the rest to you to fix them.

borngraced replied:
all done

onur-ozkan commented on mm2src/coins/z_coin/storage/walletdb/wasm/storage.rs:
```diff
@@ -0,0 +1,1494 @@
+        let tx_table = db_transaction.table::<WalletDbTransactionsTable>().await?;
+        let txs = tx_table.get_items("ticker", &ticker).await?;
+
+        let mut balance: i64 = 0;
+        for (_, note) in &maybe_notes {
+            let spent = &note.spent;
+            let value = &note.value;
+
+            for (tx_id, tx) in &txs {
+                if *tx_id == note.tx && spent.is_none() && tx.block.is_some() {
+                    let value_i64 = value.to_i64().expect("BigInt is too large to fit in an i64");
+                    balance += value_i64;
+                }
+            }
+        }
```
We should be able to achieve this without mutating `balance` using filters and maps on `maybe_notes`.

borngraced replied:
done

onur-ozkan commented on mm2src/coins/z_coin/storage/walletdb/wasm/storage.rs:
```diff
@@ -0,0 +1,1494 @@
+            }
+        }
+
+        let mut final_note = vec![];
+        for (id_note, value, note, running_sum) in &notes {
+            if let Some(_tx_block) = txs.iter().find(|(id_tx, _tx)| *id_tx == note.tx) {
+                final_note.push((id_note, value, note, running_sum));
+            }
+        }
```
Same as `maybe_notes`. These iterations/mutatings are duplicated a lot, I will leave the rest to you instead of pointing all of them.

borngraced replied:
all done

onur-ozkan commented on mm2src/coins/z_coin/storage/walletdb/wasm/storage.rs:
```diff
@@ -0,0 +1,1494 @@
+        }
+
+        let mut unspent_notes = vec![];
+        for (id, value, note, sum) in &final_note {
+            if **sum < i64::from(target_value) {
+                unspent_notes.push((*id, *value, *note, *sum))
+            }
+        }
+
+        if let Some(note) = final_note.iter().find(|(_, _, _, sum)| **sum >= target_value.into()) {
```
can we handle both of the operations in a single loop?

borngraced replied:
not really. do you have an idea?

onur-ozkan replied:
```rs
for (id, value, note, sum) in &final_note {
    if **sum < i64::from(target_value) || **sum >= target_value.into() {
        unspent_notes.push((*id, *value, *note, *sum))
    }
}
```

Can something like above do the work?

borngraced replied:
no it can't work.
the first loop selects all `notes` that their `values` are lesser then the `target_value`, while the second loop select only the first `notes` which it's value is`>=` `target_height`

onur-ozkan reviewed (commented):
next review iteration:

borngraced commented on mm2src/coins/z_coin/storage/walletdb/wasm/tables.rs:
```diff
@@ -69,53 +75,68 @@ impl TableSignature for WalletDbBlocksTable {
-    memo: String,
-    spent: BeBigUint,
-    ticker: String,
+    // references transactions(id_tx)
+    pub tx: u32,
+    pub output_index: u32,
+    // references accounts(account)
+    pub account: BigInt,
+    pub diversifier: Vec<u8>,
+    pub value: BigInt,
+    pub rcm: Vec<u8>,
+    pub nf: Option<Vec<u8>>, // unique
+    pub is_change: Option<bool>,
+    pub memo: Option<Vec<u8>>,
+    // references transactions(id_tx)
+    pub spent: Option<BigInt>,
```
done, thanks.

borngraced commented on mm2src/coins/z_coin/storage/walletdb/wasm/storage.rs:
```diff
@@ -0,0 +1,1494 @@
+pub type WalletDbRes<T> = MmResult<T, ZcoinStorageError>;
+pub type WalletDbInnerLocked<'a> = DbLocked<'a, WalletDbInner>;
+
+#[macro_export]
+macro_rules! num_to_bigint {
```
ty!

borngraced commented on mm2src/coins/z_coin/storage/walletdb/wasm/storage.rs:
```diff
@@ -0,0 +1,1494 @@
+    ) -> MmResult<Self, ZcoinStorageError> {
+        let ticker = builder.ticker;
+        let consensus_params = builder.protocol_info.consensus_params.clone();
+        let db = WalletIndexedDb::new(builder.ctx, ticker, consensus_params).await?;
+        let evk = ExtendedFullViewingKey::from(z_spending_key);
+        let extrema = db.block_height_extrema().await?;
+        let min_sync_height = extrema.map(|(min, _)| u32::from(min));
+        let init_block_height = checkpoint_block.clone().map(|block| block.height);
+        let get_evk = db.get_extended_full_viewing_keys().await?;
```
great suggestion

borngraced commented on mm2src/coins/z_coin/storage/walletdb/wasm/storage.rs:
```diff
@@ -0,0 +1,1494 @@
+        for (account, extfvk) in extfvks.iter().enumerate() {
+            let account_int = num_to_bigint!(account)?;
+            // address
+            let address = extfvk.default_address().unwrap().1;
```
I suppose.. same thing in done in librustzcash

onur-ozkan commented on mm2src/coins/z_coin/storage/walletdb/wasm/storage.rs:
```diff
@@ -0,0 +1,1494 @@
+
+#[async_trait]
+impl DbInstance for WalletDbInner {
+    fn db_name() -> &'static str { DB_NAME }
```
Thanks to @laruh, she did it already. This is no longer needed.

onur-ozkan commented on mm2src/coins/z_coin/storage/walletdb/wasm/storage.rs:
```diff
@@ -0,0 +1,1471 @@
+        &self,
+        extfvks: &[ExtendedFullViewingKey],
+    ) -> MmResult<(), ZcoinStorageError> {
+        let ticker = self.ticker.clone();
```
Unnecessary clone which causes double cloning later.

onur-ozkan reviewed (commented):
Last review iteration, looks fine to me other than that

shamardy commented on mm2src/coins/z_coin/storage/blockdb/blockdb_sql_storage.rs:
```diff
@@ -0,0 +1,219 @@
+
+impl BlockDbImpl {
+    #[cfg(all(not(test)))]
+    pub async fn new(_ctx: MmArc, ticker: String, path: Option<impl AsRef<Path>>) -> MmResult<Self, ZcoinStorageError> {
```
why does path have to be an option?

borngraced replied:
it's not needed for WASM only sqlite

shamardy replied:
But I see that `None` is never used, the path always `Some` in all uses of this function.

borngraced replied:
refactored.

shamardy commented on mm2src/coins/z_coin/z_rpc.rs:
```diff
@@ -174,17 +160,21 @@ impl ZRpcOps for LightRpcClient {
         while let Some(block) = Pin::new(&mut response).get_mut().message().await? {
             debug!("Got block {:?}", block);
-            on_block(block)?;
+            let height = u32::try_from(block.height)
+                .map_err(|_| UpdateBlocksCacheErr::InternalError("Block height too large".to_string()))?;
```
`block.height` is retrieved from Grpc server, so if it's larger than `u32::MAX`, it's because the server sent a larger number and the error is not due to internal problem, in this case the error should probably be `DecodeError`.

borngraced replied:
thanks. done

shamardy commented on mm2src/coins/z_coin/storage/blockdb/blockdb_sql_storage.rs:
```diff
@@ -0,0 +1,219 @@
+    }
+
+    pub(crate) async fn get_latest_block(&self) -> MmResult<u32, ZcashClientError> {
+        Ok(query_single_row(
+            &self.db.lock().unwrap(),
+            "SELECT height FROM compactblocks ORDER BY height DESC LIMIT 1",
+            [],
+            |row| row.get(0),
+        )?
+        .unwrap_or(0))
```
Shouldn't we use `async_blocking` here?

borngraced replied:
yeah done.

shamardy commented on mm2src/coins/z_coin/storage/blockdb/blockdb_sql_storage.rs:
```diff
@@ -0,0 +1,219 @@
+    }
+
+    pub(crate) async fn rewind_to_height(&self, height: u32) -> MmResult<usize, ZcoinStorageError> {
+        self.db
+            .lock()
+            .unwrap()
+            .execute("DELETE from compactblocks WHERE height > ?1", [height])
+            .map_to_mm(|err| ZcoinStorageError::RemoveFromStorageErr(err.to_string()))
```
Same here regarding `async_blocking`

shamardy commented on mm2src/coins/z_coin/storage/blockdb/blockdb_sql_storage.rs:
```diff
@@ -0,0 +1,219 @@
+    }
+
+    pub(crate) async fn get_earliest_block(&self) -> Result<u32, ZcashClientError> {
+        Ok(query_single_row(
+            &self.db.lock().unwrap(),
+            "SELECT MIN(height) from compactblocks",
+            [],
+            |row| row.get::<_, Option<u32>>(0),
+        )?
+        .flatten()
+        .unwrap_or(0))
```
and here

shamardy reviewed (commented):
Huge work! First review iteration from my side!

borngraced commented on mm2src/mm2_db/src/indexed_db/indexed_db.rs:
```diff
@@ -48,8 +48,8 @@ pub use db_driver::{DbTransactionError, DbTransactionResult, DbUpgrader, InitDbE
 use db_driver::{IdbDatabaseBuilder, IdbDatabaseImpl, IdbObjectStoreImpl, IdbTransactionImpl, OnUpgradeNeededCb};
-use indexed_cursor::{cursor_event_loop, CursorBuilder, CursorDriver, CursorError, CursorFilters, CursorResult,
-                     DbCursorEventTx};
+pub use indexed_cursor::CursorError;
```
thanks. done

shamardy replied:
Please keep `indexed_cursor` imports to one place, there is 2 lines for it here
https://github.com/KomodoPlatform/komodo-defi-framework/blob/9ac0af8c479036ae49ee0078b933f33bf5790cd4/mm2src/mm2_db/src/indexed_db/indexed_db.rs#L51-L52

borngraced commented on mm2src/coins/lp_coins.rs:
```diff
@@ -32,6 +32,7 @@
 #![feature(hash_raw_entry)]
 #![feature(stmt_expr_attributes)]
 #![feature(result_flattening)]
+#![feature(let_chains)]
```
nope..removed

shamardy commented on mm2src/coins/z_coin/storage.rs:
```diff
@@ -1,5 +1,201 @@
+
+#[derive(Clone)]
+pub struct DataConnStmtCacheWrapper {
+    #[cfg(not(target_arch = "wasm32"))]
+    pub cache: DataConnStmtCacheAsync<ZcoinConsensusParams>,
+    #[cfg(target_arch = "wasm32")]
+    pub cache: DataConnStmtCacheWasm,
```
These 2 fields don't need to be public, accessing `cache` is the job of the implemented `inner()` function below.

shamardy commented on mm2src/coins/z_coin/storage.rs:
```diff
@@ -1,5 +1,201 @@
+    pub fn new(cache: DataConnStmtCacheAsync<ZcoinConsensusParams>) -> Self { Self { cache } }
+    #[cfg(target_arch = "wasm32")]
+    pub fn new(cache: DataConnStmtCacheWasm) -> Self { Self { cache } }
+    #[cfg(not(target_arch = "wasm32"))]
+    pub fn inner(&self) -> DataConnStmtCacheAsync<ZcoinConsensusParams> { self.clone().cache }
+    #[cfg(target_arch = "wasm32")]
+    pub fn inner(&self) -> DataConnStmtCacheWasm { self.clone().cache }
```
The inner function pattern should return a reference. The functions here should be inlined as well.

shamardy commented on mm2src/coins/z_coin/storage.rs:
```diff
@@ -1,5 +1,201 @@
+    Ok(())
+}
+
+/// Scans at most `limit` new blocks added to the cache for any transactions received by
+/// the tracked accounts.
```
`limit` is not used anymore, please update doc comments.

borngraced replied:
done

shamardy reviewed (commented):
Next review iteration! This PR will require a few more iterations after this one.

borngraced commented on mm2src/coins/z_coin/storage/walletdb/wallet_sql_storage.rs:
```diff
@@ -0,0 +1,119 @@
+    evk: ExtendedFullViewingKey,
+    continue_from_prev_sync: bool,
+) -> Result<WalletDbAsync<ZcoinConsensusParams>, MmError<ZcoinClientInitError>> {
+    let db = WalletDbAsync::for_path(wallet_db_path, consensus_params)
+        .map_to_mm(|err| ZcoinClientInitError::ZcashDBError(err.to_string()))?;
+
+    run_optimization_pragmas_helper(&db)?;
```
okay

shamardy replied:
I think opening the connection can be blocking too, maybe we should use async_blocking here too https://github.com/KomodoPlatform/komodo-defi-framework/blob/c1ed8cbb15279c879dde7203ccdcc61f8716dbab/mm2src/coins/z_coin/storage/walletdb/wallet_sql_storage.rs#L27-L28

shamardy replied:
Another instance of opening connection https://github.com/KomodoPlatform/komodo-defi-framework/blob/c1ed8cbb15279c879dde7203ccdcc61f8716dbab/mm2src/coins/z_coin/storage/blockdb/blockdb_sql_storage.rs#L47

borngraced replied:
all done.

shamardy commented:
@borngraced please fix clippy problems

shamardy commented on mm2src/coins/z_coin/storage/walletdb/wasm/mod.rs:
```diff
@@ -0,0 +1,1125 @@
+
+    fn network() -> Network { Network::TestNetwork }
+
+    //    #[wasm_bindgen_test]
```
Why is this commented out?

borngraced replied:
one of the transaction test that takes time to run

shamardy replied:
can you please leave a comment about that so that no one deletes it in the future.

shamardy commented on mm2src/coins/z_coin/storage/walletdb/wasm/storage.rs:
```diff
@@ -0,0 +1,1470 @@
+const DB_NAME: &str = "wallet_db_cache";
+const DB_VERSION: u32 = 1;
+
+pub type WalletDbRes<T> = MmResult<T, ZcoinStorageError>;
```
Isn't this the same type as `BlockDbRes`?

shamardy commented on mm2src/coins/z_coin/storage/walletdb/wasm/storage.rs:
```diff
@@ -0,0 +1,1470 @@
+    };
+}
+
+impl<'a> WalletDbShared {
+    pub async fn new(
+        builder: ZCoinBuilder<'a>,
+        checkpoint_block: Option<CheckPointBlockInfo>,
+        z_spending_key: &ExtendedSpendingKey,
+        continue_from_prev_sync: bool,
+    ) -> MmResult<Self, ZcoinStorageError> {
```
Can't this be combined with the sql implementation?

borngraced replied:
moved

shamardy replied:
We can leave it as is for now. It's just harder to maintain this way.

borngraced commented on mm2src/coins/z_coin/storage/walletdb/wasm/mod.rs:
```diff
@@ -0,0 +1,1125 @@
+        }
+    }
+
+    //        #[wasm_bindgen_test]
```
same as above

shamardy replied:
Same as above :)

borngraced commented on mm2src/coins/z_coin/z_coin_errors.rs:
```diff
@@ -291,3 +292,184 @@ pub enum SpendableNotesError {
+    NotSupported(String),
+}
+
+#[cfg(target_arch = "wasm32")]
+use mm2_db::indexed_db::cursor_prelude::*;
+#[cfg(target_arch = "wasm32")]
+use mm2_db::indexed_db::{DbTransactionError, InitDbError};
+use mm2_err_handle::mm_error::MmError;
+#[cfg(target_arch = "wasm32")] use zcash_extras::NoteId;
```
moved

shamardy reviewed (approved):
Thanks a lot for your patience with the long review process! Please fix the 2 minor comments above.

shamardy commented:
We will merge https://github.com/KomodoPlatform/komodo-defi-framework/pull/1996 to this PR first to avoid huge conflicts, QA and sec-review will be done in this branch once https://github.com/KomodoPlatform/komodo-defi-framework/pull/1996 is merged. I shall start https://github.com/KomodoPlatform/komodo-defi-framework/pull/1996 review tommorow.

laruh commented:
Hi @borngraced could you fix conflicts?

shamardy commented:
@borngraced can you please fix merge conflicts here?

shamardy commented:
@borngraced can you please provide @KomodoPlatform/qa with any changes (if there are any) between how zcoin works in native and wasm so that they can test this PR. From my side, I think both modes (native and wasm) should be tested since there are some changes in `librustzcash` library and `sqlite` functions.

Also please provide sec-review team details about deps updates or additions and the reason for each one.

borngraced commented:
### Test Cases:
Activate normally as you would do with light client cc @smk762 @kivqa 

### Sample Coin Config
```json
{
  "coin": "ARRR",
  "asset": "PIRATE",
  "fname": "Pirate",
  "txversion": 4,
  "overwintered": 1,
  "mm2": 1,
  "avg_blocktime": 1,
  "protocol": {
    "type": "ZHTLC",
    "protocol_data": {
      "consensus_params": {
        "overwinter_activation_height": 152855,
        "sapling_activation_height": 152855,
        "blossom_activation_height": null,
        "heartwood_activation_height": null,
        "canopy_activation_height": null,
        "coin_type": 133,
        "hrp_sapling_extended_spending_key": "secret-extended-key-main",
        "hrp_sapling_extended_full_viewing_key": "zxviews",
        "hrp_sapling_payment_address": "zs",
        "b58_pubkey_address_prefix": [
          28,
          184
        ],
        "b58_script_address_prefix": [
          28,
          189
        ]
      }
    }
  },
  "required_confirmations": 3
}
```

### Sample Activation Config:
```json
{
  "userpass": "testpsw",
  "method": "task::enable_z_coin::init",
  "mmrpc": "2.0",
  "params": {
    "ticker": "ARRR",
    "activation_params": {
      "mode": {
        "rpc": "Light",
        "rpc_data": {
          "electrum_servers": [
            {
              "url": "electrum3.cipig.net:30008",
              "protocol": "WSS"
            }
          ],
          "light_wallet_d_servers": [
            "http://pirate.battlefield.earth:8581"
          ]
        }
      }
    }
  }
}
```

smk762 commented:
playground build for this at https://github.com/gcharang/react-komodefi-wasm/pull/11

smk762 commented:
I've had some troubles with this but managed to get it working today. 
There seems to be an issue with https connection to the server running the lightwalletd and envoy proxy, but its ok with plain old http, e.g.
```


{
  "userpass": "testpsw",
  "method": "task::enable_z_coin::init",
  "mmrpc": "2.0",
  "params": {
    "ticker": "ARRR",
    "activation_params": {
      "mode": {
        "rpc": "Light",
        "rpc_data": {
          "electrum_servers": [
            {
              "url": "electrum3.cipig.net:30008",
              "protocol": "WSS"
            }
          ],
          "light_wallet_d_servers": [
            "http://pirate.battlefield.earth:8581"
          ]
        }
      }
    }
  }
}
```

The cloudflare link above will return a mixed content error if attempting to use the http url, so this needs to be tested locally via https://github.com/gcharang/react-komodefi-wasm

This issue does not appear to be a problem with the PR, rather a problem with the lightwalletd/envoy configs on server. I'll investigate and resolve it later, but it should not block testing / approval of this PR.

Using default params (as above), I was able to connect and sync without any problems and in a timely manner.

When testing with the additional 

           "sync_params": {
            "date": 1672704000
          }

subsequent requests to `"task::enable_z_coin::status"` consistently returned 

  "result": {
    "status": "InProgress",
    "details": "ActivatingCoin"
  },
  
with no indication of progress. Logs indicate the process was initiated, but there is no references to activation status in logs beyond this point
  
![image](https://github.com/KomodoPlatform/komodo-defi-framework/assets/35845239/bb40274c-fe80-478f-be7c-e495a81d526c)

smk762 commented:
I've managed to get a reverse proxy for grpcs working with config below:
```
upstream api {
    server 127.0.0.1:8582;
}


server {
    listen 8581 ssl http2;
    access_log            /home/atomic/logs/grpc-access.log;
    ssl_certificate       /etc/letsencrypt/live/pirate.battlefield.earth/fullchain.pem;
    ssl_certificate_key   /etc/letsencrypt/live/pirate.battlefield.earth/privkey.pem;


    location / {
        grpc_pass grpc://api;
    }
}
```
Try `https://pirate.battlefield.earth:8581` as the `light_wallet_d_server` on https://ci-builder.react-atomicdex-wasm.pages.dev/ and it should work now.

smk762 commented:
Still seems to be stuck on `ActivatingCoin` when using sync params date, but here's some additional console logs which may be related.
![image](https://github.com/KomodoPlatform/komodo-defi-framework/assets/35845239/391d9350-155e-4917-9325-2cc0318eb96c)

borngraced commented:
> Still seems to be stuck on `ActivatingCoin` when using sync params date, but here's some additional console logs which may be related. ![image](https://private-user-images.githubusercontent.com/35845239/298268346-391d9350-155e-4917-9325-2cc0318eb96c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDU3Mzc0NjcsIm5iZiI6MTcwNTczNzE2NywicGF0aCI6Ii8zNTg0NTIzOS8yOTgyNjgzNDYtMzkxZDkzNTAtMTU1ZS00OTE3LTkzMjUtMmNjMDMxOGViOTZjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTIwVDA3NTI0N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTM2YzZiMmQ4ZmRkYmFlY2Y1MmQxZDNiMGIxYTE4MGYwZGU5Yzk0MDM2MGMwYzJkYzk1NjI1ZWY4ZDI0YTdjOTkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.IfaYSN6jKkZcg4GeAzaFw-IDQGO8neJmlGYbSzPSPIk)

Can you check status after a while? ...it seems mm2 is clearing already sync blocks since you used a different starting_height/date @smk762

smk762 commented:
Its still running from earlier ans still responds with 
![image](https://github.com/KomodoPlatform/komodo-defi-framework/assets/35845239/3a40e5ea-4e4d-4ad5-a345-1bd9381f6d87)

Looking back, these are logs from time of activation, so its been almost an hour so far.
![image](https://github.com/KomodoPlatform/komodo-defi-framework/assets/35845239/c71d2a92-200f-4cd5-b154-7d92ce3e36e0)

borngraced commented:
> Its still running from earlier ans still responds with ![image](https://private-user-images.githubusercontent.com/35845239/298270235-3a40e5ea-4e4d-4ad5-a345-1bd9381f6d87.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDU3Mzg5ODEsIm5iZiI6MTcwNTczODY4MSwicGF0aCI6Ii8zNTg0NTIzOS8yOTgyNzAyMzUtM2E0MGU1ZWEtNGU0ZC00YWQ1LWEzNDUtMWJkOTM4MWY2ZDg3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTIwVDA4MTgwMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTkzMWE2OTkwZTEyZDNlNGJiMTRmZTk5NDkyMzM3MmEwYjg3YzQxOTE1OTUxOGRhZjdlMTZjMzY1ZTA3M2JlYWImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.LLLJJjJ9L0XSTHBzKR9NH5K1dIbE4zEMJEkrmWVg0cI)
> 
> Looking back, these are logs from time of activation, so its been almost an hour so far. ![image](https://private-user-images.githubusercontent.com/35845239/298270326-c71d2a92-200f-4cd5-b154-7d92ce3e36e0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDU3Mzg5ODEsIm5iZiI6MTcwNTczODY4MSwicGF0aCI6Ii8zNTg0NTIzOS8yOTgyNzAzMjYtYzcxZDJhOTItMjAwZi00Y2Q1LWIxNTQtN2Q5MmNlM2UzNmUwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTIwVDA4MTgwMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTYyNmEzYmVhZGMwMTFkOTkwZjFjMjQzOTBmNWM0OWVhMDcyZjA3N2ViYjI5MDRhZGFjMGZjOWVhNzFhMjIyZDImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.m7XSgrTxb8DYYD5zNuQEL9TRE1om9FUYmTfVMtYozKc)

ok ok.. I'll debug later today. thank you!

borngraced commented:
Apparently, there's a deadlock bug when rewinding blocks db which I have pushed a fix for.

@smk762 you can retest again please :)

Note: Sometimes it takes time for lightwalletd server to respond with blocks so you might keep seeing `InProgress` status for a short period of time 
```json
 "result": {
   "status": "InProgress",
   "details": "ActivatingCoin"
 },
```

borngraced commented:
It looks like the issue is with the stream request. When you ask for a range of blocks from a long time ago, the request waits until it gets the first message. If the starting height is very old, it can cause delays.

I need time to investigate and find a fix. Another option is to proxy lightwalletd through a websocket connection, but that would require changes to the current transport layer I implemented. 😬

cc. @smk762 @shamardy

shamardy commented on mm2src/coins/z_coin/z_rpc.rs:
```diff
@@ -684,7 +724,8 @@ impl SaplingSyncLoopHandle {
     ) -> Result<(), MmError<UpdateBlocksCacheErr>> {
         let current_block = rpc.get_block_height().await?;
-        let current_block_in_db = self.blocks_db.get_latest_block().await?;
+        let block_db = self.blocks_db.clone();
```
do we really need this clone?

borngraced replied:
yes, because mutable borrows was used here 
```rust
 if current_block >= from_block {
            let scan_blocks = rpc.scan_blocks(from_block, current_block, &block_db, self);
            scan_blocks.await?;
        }
```

shamardy replied:
Well, I would say this is not a good design and `update_blocks_cache` shouldn't be part of `SaplingSyncLoopHandle` implementation.

borngraced replied:
Ok, I'll improve it ..not sure why it was done that way before

shamardy reviewed (commented):
Thank you for the fixes! First review iteration of these fixes :)

borngraced commented on mm2src/coins/z_coin/storage/blockdb/blockdb_idb_storage.rs:
```diff
@@ -122,22 +122,26 @@ impl BlockDbImpl {
-                .with_value(i)?;
-
-            block_db.delete_item_by_unique_multi_index(index_keys).await?;
+        let mut deleted = 0;
+        let blocks = block_db.get_items("ticker", &ticker).await?;
+        for (_, block) in blocks {
+            if block.height > height {
+                block_db
+                    .delete_item_by_unique_multi_index(
+                        MultiIndex::new(BlockDbTable::TICKER_HEIGHT_INDEX)
+                            .with_value(&ticker)?
+                            .with_value(block.height)?,
+                    )
+                    .await?;
+                deleted += 1;
+            }
```
I modified this code   because requesting to delete incorrect itemId returns error so I thought to use this approach. But yeah we can use `delete_items_by_multi_index`. Seems better anyways.

borngraced replied:
done

borngraced replied:
reverted my latest changes as the formal implementation addresses the api better

shamardy replied:
Can you please elaborate why `delete_items_by_multi_index` can't be used and how current implementation addresses the api better?

borngraced replied:
it doesn't work because we are expected to delete blocks up to a specific height and not all block heights in storage

borngraced commented on mm2src/coins/z_coin/z_rpc.rs:
```diff
@@ -236,22 +271,27 @@ impl ZRpcOps for LightRpcClient {
-            handler.notify_blocks_cache_status(block.height, last_block);
+        #[cfg(target_arch = "wasm32")]
+        let requests = construct_block_range_requests(start_block, last_block);
+        for request in requests {
+            let mut response = selfi
+                .get_block_range(request)
+                .await
+                .map_to_mm(UpdateBlocksCacheErr::GrpcError)?
+                .into_inner();
+            while let Some(block) = response.next().await {
+                let block = block.map_err(|_| UpdateBlocksCacheErr::DecodeError("Error getting block".to_string()))?;
+                debug!("Got block {}", block.height);
+                let height = u32::try_from(block.height)
+                    .map_err(|_| UpdateBlocksCacheErr::DecodeError("Block height too large".to_string()))?;
+                db.insert_block(height, block.encode_to_vec())
+                    .await
+                    .map_err(|err| UpdateBlocksCacheErr::ZcashDBError(err.to_string()))?;
```
done

ca333 commented:
few of the added deps are already in our deps tree (e.g. byteorder, base64, blake2b_simd, etc) - also/jfyi with rust 1.32 and newer, the standard library’s read_* and write_* methods on byte slices cover many use cases previously requiring byteorder. Furthermore, some of the deps added are significantly outdated. Plz recheck if the added deps are absolutely needed and if their given utilized version is optimal.

borngraced commented:
> few of the added deps are already in our deps tree (e.g. byteorder, base64, blake2b_simd, etc) - also/jfyi with rust 1.32 and newer, the standard library’s read_* and write_* methods on byte slices cover many use cases previously requiring byteorder. Furthermore, some of the deps added are significantly outdated. Plz recheck if the added deps are absolutely needed and if their given utilized version is optimal.

replaced `byetorder::BigEndian` with https://doc.rust-lang.org/std/primitive.u32.html#method.to_be_bytes
updated `bas364` to `0.21.7`
replace and use the same version of `pin-project` that's in our deps tree

for transaction related deps, like `blake2b_simd`, `jubjub` `ff`, I tried to use same version as used in `zcash lib` so as to not encounter any issue while working with `z_transactions`

cc. @ca333

shamardy reviewed (approved):
New changes LGTM! Dependency changes still need to be approved by sec review team.
@smk762 just a reminder that both Native and Wasm need to be tested for this PR since it involves a lot of refactoring of the native implementation.

shamardy commented on mm2src/mm2_db/src/indexed_db/drivers/cursor/cursor.rs:
```diff
@@ -317,6 +317,8 @@ impl CursorDriver {
         if matches!(item_action, CursorItemAction::Include) {
             if let Some(cursor_condition) = where_ {
                 if cursor_condition(val.clone())? {
+                    // stop iteration.
+                    self.stopped = true;
```
Can we open a PR for this fix alone? this is to merge this bug fix to dev quickly and not wait for reviews or approvals on this PR and this https://github.com/KomodoPlatform/komodo-defi-framework/pull/2066/files

smk762 commented:
@borngraced I've just encountered a previously unseen error testing on wasm:
```
{
  "mmrpc": "2.0",
  "result": {
    "status": "InProgress",
    "details": {
      "TemporaryError": "z_rpc:286] status: Unknown, message: \"tonic_client:45] http:340] Invalid request: InvalidRequest(\\\"JsValue(TypeError: network error\\\\nTypeError: network error)\\\")\", details: [], metadata: MetadataMap { headers: {} }"
    }
  },
  "id": null
}
```
Is this a concern? I'll restart and try again, just curious what might cause this.

borngraced commented:
> @borngraced I've just encountered a previously unseen error testing on wasm:
> 
> ```
> {
>   "mmrpc": "2.0",
>   "result": {
>     "status": "InProgress",
>     "details": {
>       "TemporaryError": "z_rpc:286] status: Unknown, message: \"tonic_client:45] http:340] Invalid request: InvalidRequest(\\\"JsValue(TypeError: network error\\\\nTypeError: network error)\\\")\", details: [], metadata: MetadataMap { headers: {} }"
>     }
>   },
>   "id": null
> }
> ```
> 
> Is this a concern? I'll restart and try again, just curious what might cause this.
> 
> Some context: ![image](https://private-user-images.githubusercontent.com/35845239/306273072-a33eb4a2-5124-402a-8ab7-7e2fa4f02902.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg0MzY2ODQsIm5iZiI6MTcwODQzNjM4NCwicGF0aCI6Ii8zNTg0NTIzOS8zMDYyNzMwNzItYTMzZWI0YTItNTEyNC00MDJhLThhYjctN2UyZmE0ZjAyOTAyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjIwVDEzMzk0NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJhMjQwMWJjNmM0MWZhM2NhYWI3YTUzMTU0ZTI2OTVkYWUzMTBmNTk1YmU1YjFkOTA4MmMzZDkwYTVkYTZhMGEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.77rc73tWB6kgua5SBcwfjftaDEW3rIHcSPZ2CbZW9ms)
> 
> Seems that though some calls to the grpc are successful, a few are returning a protocol error.
> 
> After clearing cache, next attempt was as expected:
> 
> ```
> {
>   "mmrpc": "2.0",
>   "result": {
>     "status": "InProgress",
>     "details": {
>       "BuildingWalletDb": {
>         "first_sync_block": {
>           "requested": 2794017,
>           "is_pre_sapling": false,
>           "actual": 2794017
>         },
>         "current_scanned_block": 2794017,
>         "latest_block": 2795458
>       }
>     }
>   },
>   "id": null
> }
> ```

this is simply a network error while fetching the request prob related to the server. I don't think it's internal

smk762 commented:
Thanks for clarifying, it is as I expected. We're on day 4 of a heatwave here and it seems hot enough to melt the copper my internet arrives on, so sounds like its a "me" problem.
If replicated by others, I'll review the server to see if any optimisations are possible.

smk762 commented:
Unfortunately I'm getting these errors pretty consistently, likely due to my location / connection (unless replicated). 
After a network failed error, subsequent attempts to activate returned 

```json
{
  "mmrpc": "2.0",
  "result": {
    "status": "Error",
    "details": {
      "error": "Error on platform coin ARRR creation: Database 'MAIN::21605444b36ec72780bdf52a5ffbc18288893664::z_compactblocks_cache' is open already",
      "error_path": "lib.z_coin_activation.z_coin.z_rpc.blockdb_idb_storage.builder",
      "error_trace": "lib:104] z_coin_activation:245] z_coin:995] z_rpc:517] blockdb_idb_storage:80] builder:166]",
      "error_type": "CoinCreationError",
      "error_data": {
        "ticker": "ARRR",
        "error": "Database 'MAIN::21605444b36ec72780bdf52a5ffbc18288893664::z_compactblocks_cache' is open already"
      }
    }
  },
  "id": null
}
```
and browser tab needs to be refreshed before another attempt is possible. 

I checked the lightwallet server, and there was no issues in logs. Browser console logs revealed that even after mm2 returns network error, downloading continues in the background.

https://github.com/KomodoPlatform/komodo-defi-framework/assets/35845239/1944a9a3-beba-44e1-93fe-5928e438cd24



Would it be possible to recover and continue rather than be forced to start again from the beginning? Or stopping d/l and closing db connection for second attempt? If not, perhaps an error message indicating that a reload is required before trying again.

smk762 commented:
Perhaps not an issue, but I noticed that the response for zhtlc coin initialization once a coin has been activated remains static, i.e. if balance changes after activation completes, the balance at time of activation is still returned from the `task::enable_z_coin::status` request. This also applies to the returned `current_block` value.

It would be nice if the status returned live values for balance and block_height, though this would likely also apply to other coins so beyond the scope of this PR (I'll convert to issue).

Regression testing for TCP ZHTLC activation with a variety of optional param combinations was otherwise successful, with no defects found and no issues sending (with/without memo) or trading ZHTLC coins (as maker and taker).

Wasm tests still in progress

smk762 commented:
Given my issues with connection breaks for longer scans, I focused on the date/block param tests to (hopefully) have a connection stable long enough to activate. This was successful on first try, but the second one got stuck on activating, though console seemed to suggest it has completed. If clearing cache, I was once again able to activate. 
Refreshing the page without clearing cache and trying again, I was stuck on once again activating.

Video below at 3x speed to keep filesize low:

[forever-activating.webm](https://github.com/KomodoPlatform/komodo-defi-framework/assets/35845239/1234dae6-50d0-4950-b62f-d6660fdbdd00)

This was the third attempt of sequential leave cache / clear cache test. Note that the sync param was changed after the first successful attempt and the failed 2nd attempts before clearing cache. When the sync param does not change between attempts, activation completes successfully.

borngraced commented:
> Given my issues with connection breaks for longer scans, I focused on the date/block param tests to (hopefully) have a connection stable long enough to activate. This was successful on first try, but the second one got stuck on activating, though console seemed to suggest it has completed. If clearing cache, I was once again able to activate. Refreshing the page without clearing cache and trying again, I was stuck on once again activating.
> 
> Video below at 3x speed to keep filesize low:
> 
>  forever-activating.webm 
> 
> This was the third attempt of sequential leave cache / clear cache test. Note that the sync param was changed after the first successful attempt and the failed 2nd attempts before clearing cache. When the sync param does not change between attempts, activation completes successfully.
> 
> Retesting this case on TCP, it was fine on the second activation with new sync params without needing to `rm MM2.db`

thanks for the well organized bug report 😄 .

can you try again now, I believe it shouldn't stuck anymore when you cancel and try to re-activate coin again without refreshing or clearing cache. @smk762

smk762 commented:
>  I believe it shouldn't stuck anymore when you cancel and try to re-activate coin again without refreshing or clearing cache. 

Great work, thanks! Confirm I no longer faced issues attempting to activate zhtlc with wasm after changing sync params and not clearing cache.

borngraced commented:
> Unfortunately I've been unable to perform withdrawals with ZHTLC coins in WASM. In CLI, the https://komodo-docs-revamp-2023.pages.dev/en/docs/atomicdex/api/v20-dev/task_withdraw/ methods are used, but in wasm, this returns
> 
> ```
> {
>   "mmrpc": "2.0",
>   "result": {
>     "status": "Error",
>     "details": {
>       "error": "'ARRR' coin doesn't support 'init_withdraw' yet. Consider using 'withdraw' request instead",
>       "error_path": "init_withdraw",
>       "error_trace": "init_withdraw:137]",
>       "error_type": "CoinDoesntSupportInitWithdraw",
>       "error_data": {
>         "coin": "ARRR"
>       }
>     }
>   },
>   "id": 0
> }
> ```
> 
> I also tried the v2 withdraw method, which returned
> 
> ```
> {
>   "mmrpc": "2.0",
>   "error": "Internal error: Zcoin doesn't support legacy withdraw",
>   "error_path": "z_coin",
>   "error_trace": "z_coin:1737]",
>   "error_type": "InternalError",
>   "error_data": "Zcoin doesn't support legacy withdraw",
>   "id": 0
> }
> ```
> 
> and the v1 withdraw method returns
> 
> ```
> {
>   "error": "rpc:156] rpc:215] dispatcher_legacy:141] dispatcher_legacy:151] z_coin:1737] Internal error: Zcoin doesn't support legacy withdraw"
> }
> ```

@smk762, after the build is complete, you can use the roc commands below to initiate a withdrawal
`task::withdraw::init`
`task::withdraw::status`
`task::withdraw::cancel`

smk762 commented:
I loaded up `b28e3fa` in wasm playground and tried to withdraw again, but now it is (incorrectly) telling me I'm too poor.


https://github.com/KomodoPlatform/komodo-defi-framework/assets/35845239/bb46cc15-a0c2-4e84-af54-de086bd7033f

borngraced commented:
> I loaded up `b28e3fa` in wasm playground and tried to withdraw again, but now it is (incorrectly) telling me I'm too poor.
> 
>  cant-spend.mp4

@smk762 withdraw has been enabled and tested, you can try withdrawing now without any issue
